### DISCUSSION TITLE
How to type async higher-order resolver in Typescript?

### DISCUSSION THREAD

**Discussion by nik-webdevelop** [mswjs/msw#1802] [Q&A]
How can I type this higher-order resolver? I keep getting different errors no matter what approach I use

```
import { delay } from 'msw';

export const withDelay =
  (resolver) =>
  async (...args) => {
    if (import.meta.env.MODE !== 'test') await delay();

    return resolver(...args);
  };
```

msw 2.0.0
ts 4.9.3

**Comment by kettanaito** [comment]
Hi, @nik-webdevelop. I gave it a try [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgCxjMAaOBzKBDMZARwBssATAUxLwE8sAlSgZ0gDtnKnmISA3SlCwARSgDM8AVxIwAQhHK0AKrTCU4AXzhioEEHADkIZgHcDAKHOUAHpFjbJbAMYxgENnBPAYyUTVoAPACi1jD4cDYwlGzkzHBMTtDkAcxhwGzYWI4A1mwQJmwAfIyURJIscgrKquqR0bFwohLSlYoqaiWs7pzybTURofVxTVIyvdVqhQAUUCy8AlAAXPEs7Jzc84LBofglZRXj7ZSda5SHNYUAlMvcpxv8WyFheHvlqecdK10cZ1VHhYhzHA4LMYJIoB48MxaM44FM8FBsMxLnAALwAhBA4FwPAmPDeOBUfxTACMAAYKZcscDQeCPLMeA8oPDEcisRpzByrLZoPBEhx4Mg8DESII4qi4ABtLGodAAOkgqQCSGA5GWqSg6WwmiwSFm+1SywMACMqgZNNMDAB6BkQcFOShWxaqgxYLw+Px0KZTPWlN4wLBgBF4YyaFHowHYuBB-DGOWq6kgyhgiFwdmXS4YLG4AjEEhy-ZQQJsSgLXVwVXqtIZC1TAwAcWTAFVOFBXZ5vL5qF6fXA+AjgHhjaK4hpwxjE-3NUOR-HyInaRD01SALrmIA) but soon learned we need to ship better public-facing types to make higher-order resolvers a more pleasant experience. Thanks for making this surface.

**Reply by nik-webdevelop** [reply]
@kettanaito thanks for such a quick response! I guess I will do some typecastings for now and looking forward to the type updates. Thank you

**Reply by kettanaito** [reply]
I need to think of a better way to achieve this. It's a bit tangled since the response resolver infers data from the handler. But you can see that inference working even in my playground—we don't have to expose any types to support it, TypeScript is extremely good at this. 

It's mainly the type incompatibility of the `<R>(resolver: R): R` that bewilders me.

**Comment by singhAmandeep007** [comment]
Here is my approach with authorization validation. 


```ts
// middleware.ts

import { DefaultBodyType, HttpResponse, HttpResponseResolver, PathParams } from "msw";

export function withAuth<RequestBodyType extends DefaultBodyType, ResponseBodyType extends DefaultBodyType>(
  resolver: HttpResponseResolver
): HttpResponseResolver<PathParams, RequestBodyType, ResponseBodyType> {
  // @ts-expect-error - ignore type checking for now
  return async (args) => {
    const { request } = args;

    // get Authorization header
    const authHeader = request.headers.get("Authorization");

    // check if Authorization header exists
    if (!authHeader) {
      return new HttpResponse(null, { status: 401 });
    }

    // validate Bearer token format
    const [scheme, token] = authHeader.split(" ");
    if (scheme !== "Bearer" || !token) {
      return new HttpResponse(
        JSON.stringify({
          error: "Invalid token format",
          message: "Authorization header must be in the format: Bearer <token>",
        }),
        {
          status: 401,
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
    }

    return resolver(args);
  };
}


```


```ts
// posts.ts

import { http, HttpResponse } from "msw";

import { faker } from "@faker-js/faker/locale/en";

export const getPosts = http.get(
  "/posts",
  withAuth(({ request, cookies }) => {
    const generatePosts = () => {
      const posts = [];
      for (let i = 0; i < 10; i++) {
        posts.push({
          id: faker.string.uuid(),
          title: faker.lorem.sentence(),
          userId: faker.string.uuid(),
        });
      }
      return posts;
    };

    return HttpResponse.json(generatePosts());
  })
);

```
